airbyte-integrations/singer/base-singer/base_singer/singer_helpers.py

```diff
@@ -37,27 +19,58 @@ def to_json(string):
         return False
 
 
+def is_field_metadata(metadata):
+    if len(metadata.get("breadcrumb")) != 2:
+        return False
+    else:
+        return metadata.get("breadcrumb")[0] != "property"
+
+
+@dataclass
+class Catalogs:
+    singer_catalog: object
+    airbyte_catalog: AirbyteCatalog
+
+
 class SingerHelper:
     @staticmethod
     def spec_from_file(spec_path) -> AirbyteSpec:
         with open(spec_path) as file:
             spec_text = file.read()
         return AirbyteSpec(spec_text)
 
-    # todo: support stderr in the discover process
     @staticmethod
-    def discover(shell_command, transform=(lambda x: AirbyteSchema(x))) -> AirbyteSchema:
+    def get_catalogs(logger, shell_command, singer_transform=(lambda catalog: catalog), airbyte_transform=(lambda catalog: catalog)) -> Catalogs:
         completed_process = subprocess.run(shell_command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
                                            universal_newlines=True)
-        return transform(completed_process.stdout)
+
+        for line in completed_process.stderr.splitlines():
+            logger(line, "ERROR")
+
+        airbyte_streams = []
+        singer_catalog = singer_transform(json.loads(completed_process.stdout))
```

Comment:
Should we only try to load the line if the line starts with `{` so we don't have the risk of reading some debug logs?

Reply:
I think most singer taps write all logs to `stderr` for the discover command and the entire `stdout` is expected to be readable as the catalog. I think it's probably fine to expect that behavior for now and adapt if we run into a case that violates that expectation. I imagine for that case we should be doing custom work at that source level to handle it.